lib/puppet/provider/mysql_user/mysql.rb

```diff
@@ -23,6 +23,13 @@ def self.instances
       # rubocop:enable Layout/LineLength
       @max_user_connections, @max_connections_per_hour, @max_queries_per_hour, @max_updates_per_hour, ssl_type, ssl_cipher,
       x509_issuer, x509_subject, @password, @plugin, @authentication_string = mysql_caller(query, 'regular').chomp.split(%r{\t})
+
+      if @plugin == 'caching_sha2_password'
+        # Escaping all single quotes to prevent errors when password generated it
+        @password = @password.gsub("'") { "\\'" }
+        @password = '0x' + @password.each_byte.map { |b| b.to_s(16) }.join
```

Comment:
We've encountered idempotency problem with this line. The hash displayed by Puppet using `puppet resource mysql_user` doesn't match what is given by `SHOW CREATE USER`:

* the string given by `show create user` is all uppercase but in puppet it is all lowercase
* some `0` are not present when using puppet resource, I believe this is because the `to_s(16)` doesn't account for bytes lower than 16

I've been able to fix these two problems with the following code:

```suggestion
        @password = '0x' + @password.each_byte.map { |b| '%02X' % b.to_i }.join
```